Cfml support

scottsteinbeck opened the pull request:
Added support for CFML - ColdFusion/Lucee support for the curl converter site.

verhovsky commented on test/test-utils.ts:
```diff
@@ -40,6 +41,11 @@ const converters = {
   },
+  cfml: {
+    name: "ColdFusion",
+    extension: ".cfm",
```
You have .cfm here but the test files are .js

scottsteinbeck replied:
I will update them. I chose js as the extension just due to the color highlighting working better for the script syntax.

scottsteinbeck commented:
@verhovsky, if you had a chance, could you take a look at the updated PR, Thanks!

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,113 @@
+  let cfmlCode = "";
+
+  cfmlCode += "httpService = new http();\n";
+  cfmlCode += 'httpService.setUrl("' + request.url + '");\n';
```
You should quote anything because there's nothing stopping someone from passing in a URL with a quote in it to curl, even if that doesn't make sense.

```suggestion
  cfmlCode += 'httpService.setUrl("' + quote(request.url) + '");\n';
```

scottsteinbeck replied:
good point i will wrap them

scottsteinbeck replied:
Updated PR with `quote`

verhovsky commented on README.md:
````diff
@@ -13,7 +13,7 @@ data = {
 ```
 
-You can choose the output language by passing `--language <language>`. The options are `python` (the default), `javascript`/`node`/`node-request`, `php`, `go`, `java`, `r`, `rust`, `elixir`, `dart`, `matlab` and a few more.
+You can choose the output language by passing `--language <language>`. The options are `python` (the default), `javascript`/`node`/`node-request`, `php`, `go`, `java`, `r`, `rust`, `elixir`, `dart`, `matlab`,`cfml` and a few more.
````
```suggestion
You can choose the output language by passing `--language <language>`. The options are `python` (the default), `javascript`/`node`/`node-request`, `php`, `go`, `java`, `r`, `rust`, `elixir`, `dart`, `matlab`, `cfml` and a few more.
```

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,113 @@
+        '", value="' +
+        quote(headerValue) +
+        '");\n';
+    }
```
```suggestion
    }
    deleteHeader(request, 'cookie');
```

you'll also need to import that up top

scottsteinbeck replied:
Added into the new PR

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,113 @@
+  if (request.cookies) {
+    for (const [headerName, headerValue] of request.cookies) {
+      cfmlCode +=
+        'httpService.addParam(type="cookie",name="' +
```
```suggestion
        'httpService.addParam(type="cookie", name="' +
```

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,113 @@
+      }
+    } else if (
+      request.isDataBinary &&
+      (request.data as string).charAt(0) === "@"
```
`--data @filename` means "read the --data from *filename*. I think since you're not generating code that reads from a filename, it's best to just leave it as-is.

scottsteinbeck replied:
Updated code to read in file contents as expected

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,113 @@
+    }
+  }
+
+  cfmlCode += "result = httpService.send().getPrefix();\n";
```
let's add an extra newline here

```suggestion
  cfmlCode += "\nresult = httpService.send().getPrefix();\n";
```

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,113 @@
+
+  cfmlCode += "httpService = new http();\n";
+  cfmlCode += 'httpService.setUrl("' + quote(request.url as string)  + '");\n';
+  cfmlCode += 'httpService.setCharset("utf-8");\n';
```
What is this and what's the reason for adding this line in general? Is this the `charset` part of the `Content-Type: text/html; charset=UTF-8` header? If so, we shouldn't add this line if `hasHeader(request, 'content-type')`. Ideally we could parse the `content-type` header and remove the `charset=UTF-8` and add this line instead if it exists, or just don't add it at all for any reason.

scottsteinbeck replied:
I'll just leave it out

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,113 @@
+  cfmlCode += 'httpService.setCharset("utf-8");\n';
+
+  cfmlCode +=
+    'httpService.setMethod("' + quote(request.method.toUpperCase()) + '");\n';
```
If you're passing it as a string, just pass it as is

```suggestion
    'httpService.setMethod("' + quote(request.method) + '");\n';
```

curl lets you do `curl -X PoSt` and it will actually send the letter case you specify.

verhovsky commented on package-lock.json:
```diff
@@ -1,7896 +0,0 @@
-{
```
this file shouldn't be deleted, you can re-generate it with `npm install`

scottsteinbeck replied:
Sorry forgot that got uploaded

verhovsky commented on package-lock.json:
```diff
@@ -1,4607 +1,8 @@
   "name": "curlconverter",
   "version": "4.0.0-alpha.10",
-  "lockfileVersion": 2,
+  "lockfileVersion": 1,
```
you're using an old version of `npm` I think

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,115 @@
+    }
+  }
+
+  if (request.hasOwnProperty("cookie")) {
```
```suggestion
  if (request.cookie) {
```

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,115 @@
+    util.deleteHeader(request, "Cookie");
+  }
+
+  if (request.headers) {
```
```suggestion
  if (request.headers && request.headers.length) {
```

since it can be an empty list now

scottsteinbeck commented:
@verhovsky Hopefully I got it all handled, trying to get this published but I don't want to keep bothering you with updates.

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,115 @@
+    const proxy = request.proxy.replace(":" + proxyPort, "");
+
+    cfmlCode += 'httpService.setProxyServer("' + quote(proxy) + '");\n';
+    cfmlCode += 'httpService.setProxyPort("' + quote(proxyPort) + '");\n';
```
```suggestion
    if (proxyPort) {
      cfmlCode += 'httpService.setProxyPort("' + quote(proxyPort) + '");\n';
    }
```

scottsteinbeck commented:
@verhovsky any chance you would have time to review the last bit of changes for me. Also, once the PR is approved do I need to fork and PR the repo for the site & vscode extension or is that something you do?

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,117 @@
+        quote(headerValue) +
+        '");\n';
+    }
+  }
+
+  if (request.cookie) {
+    util.deleteHeader(request, "Cookie");
```
```suggestion
    util.deleteHeader(request, "Cookie");
```

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,114 @@
+
+  if (request.timeout) {
+    cfmlCode +=
+      'httpService.setTimeout("' + (parseInt(request.timeout) || 0) + '");\n';
```
`httpService.setTimeout` takes a string?

scottsteinbeck replied:
in CFML it will interpret it as a number since it is a dynamically typed language, but it will accept both, i can remove the quotes since it will always be a number from curl

scottsteinbeck replied:
don't go away, I'm updating it right now, give me 2 minutes

scottsteinbeck replied:
done

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,114 @@
+  if (request.data || request.multipartUploads) {
+    if (request.multipartUploads) {
+      for (const [multipartKey, multipartValue] of request.multipartUploads) {
+        if (multipartValue.charAt(0) === "@") {
```
```suggestion
        if (!data.isDataRaw && multipartValue.charAt(0) === "@") {
```

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,114 @@
+  if (request.data || request.multipartUploads) {
+    if (request.multipartUploads) {
+      for (const [multipartKey, multipartValue] of request.multipartUploads) {
+        if (!data.isDataRaw && multipartValue.charAt(0) === "@") {
```
```suggestion
        if (multipartValue.charAt(0) === "@" && !request.isDataRaw) {
```

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  if (request.proxy) {
+    const proxyPart = (request.proxy as string).match(/:([0-9]{2,})/);
+    const proxyPort = proxyPart && proxyPart.length > 1 ? proxyPart[1] : "";
+    const proxy = request.proxy.replace(":" + proxyPort, "");
```
This will remove `:` in the URL's "`://`" when there's no port and also we might as well handle a single digit port number.

Curl will choose 1080 as the default port if one isn't provided https://curl.se/docs/manpage.html#-x
Does CFML do that too? 

```suggestion
    let proxy = request.proxy
    let proxyPort = "";
    const proxyPart = (request.proxy as string).match(/:([0-9]+)/);
    if (proxyPart) {
      proxy = request.proxy.slice(0, proxyPort.index);
      proxyPort = proxyPart[1]
    }
```

scottsteinbeck replied:
It looks like CFML will default to 80 so maybe we need to set it to 1080 since it is expected to match curl defaults

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    if (request.proxyAuth) {
+      const proxyauth = request.proxyAuth.split(":");
```
```suggestion
      const proxyauth = request.proxyAuth.split(/:(.*)/s, 2);
```

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    if (request.proxyAuth) {
+      const proxyauth = request.proxyAuth.split(":");
+      cfmlCode += 'httpService.setProxyUser("' + quote(proxyauth[0]) + '");\n';
+      cfmlCode +=
+        'httpService.setProxyPassword("' + quote(proxyauth[1]) + '");\n';
```
```suggestion
      const [proxyUser, proxyPassword] = request.proxyAuth.split(/:(.*)/s, 2);
      cfmlCode += 'httpService.setProxyUser("' + quote(proxyUser) + '");\n';
      cfmlCode +=
        'httpService.setProxyPassword("' + quote(proxyPassword || "") + '");\n';
```

scottsteinbeck commented:
@verhovsky I updated the code to use your suggestion, just modified it slightly to default the proxy port to 1080. It is stored as a string `let proxyPort = '1080'` due to the TS validation getting mad at the variable TYPE change from string to number. This should not affect anything since it is still being added to the final output as a string.

verhovsky commented:
We can't add it to the VS Code extension because it has to ship an old version, see https://github.com/curlconverter/curlconverter-vscode/issues/1. I'll add it to the site myself, but it'll be hard because we've run out of room.

<img width="1185" alt="Screen Shot 2022-04-08 at 17 09 14" src="https://user-images.githubusercontent.com/5687998/162548376-c89c79a3-e43c-4e0b-a4d5-67fdd1ae9b13.png">

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,116 @@
+  if (request.data || request.multipartUploads) {
+    if (request.multipartUploads) {
+      for (const [multipartKey, multipartValue] of request.multipartUploads) {
+        if (multipartValue.charAt(0) === "@" && !request.isDataRaw) {
```
sorry, this was a bad suggestion, it's not applicable to `-F`
```suggestion
        if (multipartValue.charAt(0) === "@") {
```

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,116 @@
+        }
+      }
+    } else if (
+      request.isDataBinary &&
```
```suggestion
      !request.isDataRaw &&
```

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,116 @@
+    }
+  }
+
+  if (request.auth) {
+    cfmlCode +=
+      'httpService.addParam(type="header", name="Authorization", value="' +
+      quote(request.auth.join(":")) +
+      '");\n';
+  }
+
```
```suggestion
```

remove this because you actually need to set it to `'Basic ' + btoa(request.auth.join(':'))`, we should add it to `request.headers` while building the request.

scottsteinbeck replied:
Actually, the HTTP tag does have a set username and password attributes, I will check what that is doing behind the scenes. It may be achieving the same end result of base64 encoding the string ahead of time and adding basic

scottsteinbeck replied:
Updated code to reflect auth parameters on the http object

scottsteinbeck commented:
would you mind checking through the rest of `cfml.ts` to give me all the remaining code feedback in one review that way i can address it without having to ping you so often

scottsteinbeck commented:
> We can't add it to the VS Code extension because it has to ship an old version, see [curlconverter/curlconverter-vscode#1](https://github.com/curlconverter/curlconverter-vscode/issues/1). I'll add it to the site myself, but it'll be hard because we've run out of room.
> 
> <img alt="Screen Shot 2022-04-08 at 17 09 14" width="1185" src="https://user-images.githubusercontent.com/5687998/162548376-c89c79a3-e43c-4e0b-a4d5-67fdd1ae9b13.png">

What if we shortened `JavaScript` to `JS`, I think that would provide enough room

verhovsky commented on src/generators/cfml.ts:
```diff
@@ -0,0 +1,116 @@
+            '");\n';
+        }
+      }
+    } else if (
+      !request.isDataRaw &&
+      (request.data as string).charAt(0) === "@"
+    ) {
+      cfmlCode +=
+        'httpService.addParam(type="body", value="#fileReadBinary(expandPath("' +
+        quote((request.data as string).substring(1)) +
+        '"))#");\n';
```
Is there a `fileRead` as opposed to `fileReadBinary` or something? We could have two cases

```suggestion
    } else if (
      !request.isDataRaw &&
      (request.data as string).charAt(0) === "@"
    ) {
      cfmlCode +=
        'httpService.addParam(type="body", value="#' + 
        (request.isDataBinary ? "fileReadBinary" : "fileRead") +
        '(expandPath("' +
        quote((request.data as string).substring(1)) +
        '"))#");\n';
```

scottsteinbeck replied:
There is a fileread but and it returns a string

verhovsky replied:
Sorry, the "but" in your comment is throwing me off a bit. This change makes sense, right?

scottsteinbeck replied:
Sorry I was supposed to just say. Yes it does exist and it returns a string

verhovsky commented:
> if we shortened JavaScript to JS

That does work but I don't like that idea, sorry. JavaScript is probably the most or second most used converter, it should be the easiest to ctrl-f for and not everyone knows the abbreviation.

verhovsky commented:
your generator is live, thanks for working on this!

scottsteinbeck commented:
Awesome! Glad it's up, can you double check the website, I think the new build is not being used. It says the cfml converter is missing
